### PR TITLE
Mash Overrides, Restructure Prep

### PR DESCRIPTION
**Pull request by GradualSyrup**
Brings in mash changes for the overrides and some skeleton for input recording. There are some debug statements for input recording still lying around and other artifacts, will be cleaned up before release.

### CODE REVIEW

File: src/training/ledge.rs [hunk 1]
```diff
@@ -73,13 +74,30 @@ pub unsafe fn force_option(module_accessor: &mut app::BattleObjectModuleAccessor
...
+    let prev_status_kind = StatusModule::prev_status_kind(module_accessor, 0);
+
+    if status_kind == *FIGHTER_STATUS_KIND_CLIFF_WAIT && prev_status_kind == *FIGHTER_STATUS_KIND_CLIFF_CATCH { // For regular ledge grabs, we were just in catch and want to buffer on this frame
+        should_buffer = (LEDGE_DELAY == 0) && (current_frame == 19) && (!flag_cliff);
+    } else if status_kind == *FIGHTER_STATUS_KIND_CLIFF_WAIT { // otherwise we're in "wait" from grabbing with lasso, so we want to buffer on frame
+        should_buffer = (LEDGE_DELAY == 0) && (current_frame == 18) && (flag_cliff);
+    } else {
+        should_buffer = false;
+    }
+    println!("Status: {}, Frame: {}, flag_cliff: {}, buffer: {}",status_kind,current_frame,flag_cliff, should_buffer);
```
Comment (asimon-1) [inline src/training/ledge.rs:89]: Remove debug print statement

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FROM jugeeya/cargo-skyline:3.0.0-dkp-beta
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply .devcontainer/Dockerfile:1]: Deleted.

File: src/training/clatter.rs [hunk 3]
```diff
@@ -42,14 +40,12 @@ pub unsafe fn handle_clatter(module_accessor: &mut BattleObjectModuleAccessor) {
...
         return;
     }
-    if !is_in_clatter(module_accessor) {
-        if WAS_IN_CLATTER_FLAG && MENU.mash_triggers.contains(MashTrigger::CLATTER) {
-            mash::buffer_menu_mash();
-        }
-        WAS_IN_CLATTER_FLAG = false;
+
+    if !is_in_clatter(module_accessor) { // don't input clatter if we're not in clatter
+        // TODO: remove? //EffectModule::remove_common(module_accessor, Hash40::new("sys_clatter"));
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply src/training/clatter.rs:45]: Removed TODO.

File: src/training/input_record.rs [hunk 4]
```diff
@@ -1,80 +1,254 @@
...
+    Player,
+    Cpu,
+    Lockout,
+    Standby,
+}
+
+use InputRecordState::*;
+use PossessionState::*;
+
+const FINAL_RECORD_MAX: usize = 150; // Maximum length for input recording sequences (capacity)
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply src/training/input_record.rs:29]: Recording Length is set to this for testing, we can determine what max size we actually want later. I agree that 5-10 seconds is probably what we'll want in the future.

File: training_mod_consts/src/lib.rs [hunk 5]
```diff
@@ -1,1687 +1,1999 @@
...
+    pub save_state_mirroring: SaveStateMirroring,
+    pub sdi_state: Direction,
+    pub sdi_strength: SdiFrequency,
+    pub shield_state: Shield,
+    pub shield_tilt: Direction,
+    pub stage_hazards: OnOff,
+    pub tech_state: TechFlags,
+    pub throw_delay: MedDelay,
+    pub throw_state: ThrowOption,
+    pub block_state: Action,
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply training_mod_consts/src/lib.rs:1284]: Renamed all of these `_state`s to `_override`s

File: src/training/mash.rs [hunk 6]
```diff
@@ -130,47 +144,79 @@ pub unsafe fn get_command_flag_cat(
...
+            buffer_menu_mash(action);
+        },
+        None => return,
     }
-
-    buffer_menu_mash();
 }
 
-unsafe fn should_buffer(module_accessor: &mut app::BattleObjectModuleAccessor) -> bool {
+unsafe fn should_buffer(module_accessor: &mut app::BattleObjectModuleAccessor) -> Option<Action> {
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply src/training/mash.rs:161]: Renamed to `get_buffered_action`

File: src/training/mash.rs [hunk 7]
```diff
@@ -130,47 +144,79 @@ pub unsafe fn get_command_flag_cat(
...
-        || (MENU.mash_triggers.contains(MashTrigger::PARRY) && is_in_parry(module_accessor))
-        || (MENU.mash_triggers.contains(MashTrigger::TUMBLE) && is_in_tumble(module_accessor))
-        || (MENU.mash_triggers.contains(MashTrigger::LANDING) && is_in_landing(module_accessor))
-        || (MENU.mash_triggers.contains(MashTrigger::TRUMP) && is_in_ledgetrump(module_accessor))
-        || (MENU.mash_triggers.contains(MashTrigger::FOOTSTOOL) && is_in_footstool(module_accessor))
-        // CLATTER handled in clatter.rs
-        // LEDGE handled in ledge.rs
-        // TECH handled in tech.rs
-        // MISTECH handled in tech.rs
+    if MENU.mash_triggers.contains(MashTrigger::HIT) && is_in_hitstun(module_accessor) {
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply src/training/mash.rs:163]: Yeah this is the best I've got at this time, but worth looking into in the future to see if there's a better way to format this.

File: src/training/ledge.rs (hunk 1 above)
Comment (GradualSyrup) [reply src/training/ledge.rs:89]: ~~don't tell me what to do~~ Deleted.

File: src/training/fast_fall.rs [hunk 8]
```diff
@@ -1,140 +1,144 @@
...
+
+/**
+ * Returns true for viable fast fall status
+ */
+fn is_correct_status(module_accessor: &mut app::BattleObjectModuleAccessor) -> bool {
+    let status;
+
+    unsafe {
+        status = StatusModule::status_kind(module_accessor);
+        if input_record::is_playback() {
```
[the remark this answers is not in the document]
Comment (GradualSyrup) [reply src/training/fast_fall.rs:86]: Unsure of final implementation for these `is_playback()` checks; will look into in the future to see how we want to finalize these guards

**Comment by asimon-1** [comment]
A few more comments:
1. The tumble override is not hooked up
2. The clatter override is not hooked up
3. The tech override is not working
4. I think we should not require the mash trigger if there's an override option set. It felt a little unintuitive to require setting an option in two places

**Comment by asimon-1** [comment]
Other TODO's

- [ ] I need to send you icons
- [ ] Airdodge mash shouldn't cause a shield flicker when grounded
- [ ] Panic when exiting quick menu
- [ ] Rename "block" to "shieldstun" to be consistent with the mash triggers menu (I think I told you to use "block" before, sorry!)

**Comment by austintraver** [comment]
This branch now includes icons for mash overrides (including fix for wrong icons)

**Comment by GradualSyrup** [comment]
This has all been worked into main via other pull requests.

**Pull request closed without merging**